crates/data_components/src/mssql.rs

```diff
@@ -0,0 +1,620 @@
+
+use crate::arrow::write::MemTable;
+
+use arrow::{
+    array::{
+        ArrayBuilder, ArrayRef, BooleanBuilder, Date32Builder, Decimal128Builder, Float32Builder,
+        Float64Builder, Int16Builder, Int32Builder, Int64Builder, NullBuilder, RecordBatch,
+        RecordBatchOptions, StringBuilder, Time64NanosecondBuilder, TimestampMillisecondBuilder,
+        TimestampNanosecondBuilder, UInt8Builder,
+    },
+    datatypes::{DataType, Date32Type, Field, Schema, SchemaRef, TimeUnit},
+};
+use datafusion::{
+    catalog::Session,
+    datasource::{TableProvider, TableType},
+    error::DataFusionError,
+    logical_expr::{Expr, TableProviderFilterPushDown},
+    physical_plan::ExecutionPlan,
+    sql::TableReference,
+};
+
+use chrono::Timelike;
+
+use std::{any::Any, sync::Arc};
+use tokio_util::compat::{Compat, TokioAsyncWriteCompatExt};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Error executing query: {source}"))]
+    QueryError { source: tiberius::error::Error },
+
+    #[snafu(display("Invalid connection string: {source}"))]
+    InvalidConnectionStringError { source: tiberius::error::Error },
+
+    #[snafu(display("Unable to connect: {source}"))]
+    SqlServerAccessError { source: tiberius::error::Error },
+
+    #[snafu(display("Unable to connect: {source}"))]
+    ConnectionPoolError {
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+
+    #[snafu(display("Failed to retrieve table schema"))]
+    SchemaRetrieval,
+
+    #[snafu(display("Unable to retrieve table schema: dataset not found"))]
+    SchemaRetrievalTableNotFound,
+
+    #[snafu(display("Unsupported data type: {data_type}"))]
+    UnsupportedType { data_type: String },
+
+    #[snafu(display("Failed to build record batch: {source}"))]
+    FailedToBuildRecordBatch { source: arrow::error::ArrowError },
+
+    #[snafu(display("No builder found for index {index}"))]
+    NoBuilderForIndex { index: usize },
+
+    #[snafu(display("Failed to downcast builder for {mssql_type}"))]
+    FailedToDowncastBuilder { mssql_type: String },
+}
+
+pub type SqlServerConnectionPool = Pool<SqlServerConnectionManager>;
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub struct SqlServerTableProvider {
+    #[allow(dead_code)]
+    conn: Arc<SqlServerConnectionPool>,
+    schema: SchemaRef,
+    table: TableReference,
+}
+
+impl SqlServerTableProvider {
+    pub async fn new(conn: Arc<SqlServerConnectionPool>, table: &TableReference) -> Result<Self> {
+        let schema = Self::get_schema(Arc::clone(&conn), table).await?;
+
+        Ok(Self {
+            conn,
+            schema,
+            table: table.clone(),
+        })
+    }
+
+    pub async fn get_schema(
+        conn: Arc<SqlServerConnectionPool>,
+        table: &TableReference,
+    ) -> Result<SchemaRef> {
+        let table_name = table.table();
+        let table_schema = table.schema().unwrap_or("dbo");
+
+        let columns_meta_query: String = format!(
```

Comment:
'dbo' - dbo (Database Owner) is the default schema when a schema is not explicitly specified. If you create or access an object without specifying the schema, SQL Server will assume it belongs to the dbo schema